File: Core/Entities/Product.cs
namespace Core.Entities
{
    public class Product : BaseEntity
    {
        public string Name { get; set; }
        public string PublicName { get; set; }
        public ProductType ProductType { get; set; }
        public int ProductTypeId { get; set; }
        public decimal Length { get; set; }
        public Material Material { get; set; }
        public int MaterialId { get; set; }
        public string Finishing { get; set; }
        public Material FinishMaterial1 { get; set; }
        public int FinishMaterial1Id { get; set; }
        public Material FinishMaterial2 { get; set; }
        public int FinishMaterial2Id { get; set; }
        public Material FinishMaterial3 { get; set; }
        public int FinishMaterial3Id { get; set; }
        public Hardware Hook1 { get; set; }
        public int Hook1Id { get; set; }
        public Hardware Hook2 { get; set; }
        public int Hook2Id { get; set; }
        public Hardware ORing1 { get; set; }
        public int ORing1Id { get; set; }
        public Hardware ORing2 { get; set; }
        public int ORing2Id { get; set; }
        public Hardware StopBar { get; set; }
        public int StopBarId { get; set; }
        public Hardware Keychain { get; set; }
        public int KeychainId { get; set; }
        public Hardware EndCaps { get; set; }
        public int EndCapsId { get; set; }
        public decimal Price { get; set; }
        public int InStock { get; set; }
        public string Description { get; set; }
        public string LastUpdated { get; set; }
        private readonly List<Photo> _photos = new List<Photo>();
        public IReadOnlyList<Photo> Photos => _photos.AsReadOnly();

        public void AddPhoto(string pictureUrl, string fileName, bool isMain = false)
        {
            var photo = new Photo
            {
                FileName = fileName,
                PictureUrl = pictureUrl
            };

            if (_photos.Count == 0) photo.IsMain = true;

            _photos.Add(photo);
        }

        public void RemovePhoto(int id)
        {
            var photo = _photos.Find(x => x.Id == id);
            _photos.Remove(photo);
        }

        public void SetMainPhoto(int id)
        {
            var currentMain = _photos.SingleOrDefault(item => item.IsMain);
            foreach (var item in _photos.Where(item => item.IsMain))
            {
                item.IsMain = false;
            }

            var photo = _photos.Find(x => x.Id == id);
            if (photo != null)
            {
                photo.IsMain = true;
                if (currentMain != null) currentMain.IsMain = false;
            }
        }



    }
}

File: client/src/app/shared/models/product.ts
export interface IProduct {
    id: number;
    name: string;
    publicName: string;
    productType: string;
    productTypeId: number;
    length: number;
    material: string;
    materialId: number;
    color: string;
    size: number;
    finishing: string;
    finishMaterial1: string;
    finishMaterial1Id: number;
    finishMaterial2: string;
    finishMaterial2Id: number;
    finishMaterial3: string;
    finishMaterial3Id: number;
    hook1: string;
    hook1Id: number;
    hook2: string;
    hook2Id: number;
    oRing1: string;
    oRing1Id: number;
    oRing2: string;
    oRing2Id: number;
    stopBar: string;
    stopBarId: number;
    keychain: string;
    keychainId: number;
    endCaps: string;
    endCapsId: number;
    price: number;
    inStock: number;
    description: string;
    lastUpdated: string;
    pictureUrl: string;
    photos: IPhoto[];
}

export interface IPhoto {
    id: number;
    pictureUrl: string;
    fileName: string;
    isMain: boolean;
}

export interface IProductToCreate {
    name: string;
    publicName: string;
    productTypeId: number;
    length: number;
    materialId: number;
    finishing: string;
    finishMaterial1Id: number;
    finishMaterial2Id: number;
    finishMaterial3Id: number;
    hook1Id: number;
    hook2Id: number;
    oRing1Id: number;
    oRing2Id: number;
    stopBarId: number;
    keychainId: number;
    endCapsId: number;
    price: number;
    inStock: number;
    description: string;
    lastUpdated: string;
    pictureUrl: string;
}

export class ProductFormValues implements IProductToCreate {
    name = '';
    publicName = '';
    productTypeId: number;
    length = 0;
    materialId: number;
    finishing = '';
    finishMaterial1Id: number;
    finishMaterial2Id: number;
    finishMaterial3Id: number;
    hook1Id: number;
    hook2Id: number;
    oRing1Id: number;
    oRing2Id: number;
    stopBarId: number;
    keychainId: number;
    endCapsId: number;
    price = 0;
    inStock = 0;
    description = '';
    pictureUrl = '';
    lastUpdated = '';

    constructor(init?: ProductFormValues) {
        Object.assign(this, init);
    }
}